### PR TITLE
Build: fixes for making Calypso run on Windows under MSYS2

### PR DESCRIPTION
**Pull request by klimeryk**
I've seen _some_ issues (#592, #1193, #2029, #2066) about running Calypso on Windows, so since I had some experience with multi-OS makefiles from my previous job, I figured I'd give this one a stab. Turns out it wasn't _that_ hard, especially when you just look at the diff :stuck_out_tongue:
The main pain point is that Node.js does not support building under Cygwin or MinGW (believe me, I've tried...), so we are stuck with a Windows version (MSYS2 comes with nodejs version 0.14 or something, so too old for our purposes). Which uses Windows-style paths. On the other hand, we need `make` - the only fresh (not 3.81 from gnuwin32 that is almost 10 years old...) one is bundled with MSYS2 or Cygwin (or babun, etc.). I've opted with MSYS2, since it's more lightweight than Cygwin. But the problem is that `make` and co. uses unix-style paths. _Furthermore_, nodejs uses `git` when fetching some dependencies, so we need to use the Windows build of git, *not* the one that can be installed in MSYS2. Argh.

Anyway, apart from these slight annoyances in the beginning, the rest is smooth sailing and everything seem to work. But that's just my initial testing, on my machine (Windows 7 64-bit). I'd love more people to test this PR, especially on Windows 8 or 10 (or whatever the cool kids use these days).

### CODE REVIEW

File: server/bundler/utils.js [hunk 1]
```diff
@@ -9,7 +9,7 @@ function getAssets( stats ) {
 			name: chunk.names[0],
 			hash: chunk.hash,
 			file: filename,
-			url: path.resolve( stats.publicPath, filename )
+			url: stats.publicPath + filename,
```
Comment (klimeryk) [inline server/bundler/utils.js]: Just a note, in case someone wonders why this change - the old version is incorrect. `path.resolve` should be used for real paths. But here `stats.publicPath` is `/calypso`, a prefix for URLs for assets. This worked on *nix systems because it looks like a Unix path, but on Windows this invocation resulted in a path like `C:\\calypso\\bundle-hash.js` which is definitely not what we want :)

File: Makefile [hunk 2]
```diff
@@ -119,9 +129,12 @@ mixedindentlint: node_modules/mixedindentlint
 	@$(RECORD_ENV) $@
 
 # generate the client-side `config` js file
-$(CLIENT_CONFIG_FILE): .env config/$(CALYPSO_ENV).json config/client.json client/config/regenerate.js
+$(CLIENT_CONFIG_FILE): .env config/$(CALYPSO_ENV).json config/client.json client/config/regenerate.js client/lib/user/shared-utils.js
 	@$(NODE) client/config/regenerate.js > $@
 
+client/lib/user/shared-utils.js: server/user-bootstrap/shared-utils.js
+	@ln -s $(realpath $<) $@
+
```
Comment (klimeryk) [inline Makefile]: Unfortunately, no version of Git for Windows that I've found [supports symlinks](http://stackoverflow.com/a/16754068/181497). So I've removed the symlink in git for `client/lib/user/shared-utils.js` and instead create it via `make`.

File: Makefile [hunk 3]
```diff
@@ -14,9 +19,9 @@ NODE_BIN := $(THIS_DIR)/node_modules/.bin
 
 # applications
 NODE ?= node
-NPM ?= $(NODE) $(shell which npm)
+NPM ?= npm
```
Comment (klimeryk) [inline Makefile]: Was there a reason why it was invoked `node npm` previously? `npm` seems like a perfectly fine standalone command. Just making sure I've not broken some exotic setup :)

Comment (blowery) [inline Makefile]: I'm not sure there was a reason for it, but previously it made sure that npm was run using the same node as the rest of the process. You could supply a NODE=/my/fancy/node as an evironment variable and have that node used. 

I don't think we ever actually used that capability though?

File: Makefile [hunk 4]
```diff
@@ -78,7 +88,7 @@ run: welcome githooks-commit install build
 	@$(NODE) build/bundle-$(CALYPSO_ENV).js
 
 node-version:
-	@NPM_GLOBAL_ROOT=$(shell $(NPM) -g root) $(BIN)/check-node-version
+	@$(BIN)/check-node-version
```
Comment (blowery) [inline Makefile]: This has to be the old way to pick up `semver` from the global install of `npm` /cc @aduth

File: bin/check-node-version [hunk 5]
```diff
@@ -1,6 +1,7 @@
 #!/usr/bin/env node
 const requiredVersion = require( '../package' ).engines.node,
-	semver = require( process.env.NPM_GLOBAL_ROOT + '/npm/node_modules/semver' );
+	path = require( 'path' ),
+	semver = require( path.join( process.env.NPM_GLOBAL_ROOT, 'semver' ) );
```
Comment (blowery) [inline bin/check-node-version]: oh, sneaky!

File: Makefile (hunk 3 above)
Comment (klimeryk) [inline Makefile]: Ah, I see. But calling `node C:\path\to\npm` fails on Windows - because `npm` is a bash script there (also available in `cmd` version, which is just as useful/less). While on Linux/OS X it's a nodejs script - go figure. But there's no guarantee that won't change, so it's best to just call `npm` directly and let `nvm`/whatever (or just the user who wants `exotic-node`) figure out the right `node`/`npm` combo. If needed, `NODE` and `NPM` variables can be provided to `make`.

File: Makefile (hunk 4 above)
Comment (klimeryk) [inline Makefile]: Another interesting tidbit - `semver` is not installed (globally, or locally, of course) on Windows as part of a standard nodejs install. So you need to explicitly install it (it's in the instructions I've written as part of this PR). I've seen a discussion about how to pull in this dependency, but nothing better than the current approach struck my fancy. And since `semver` is available (always?) on Linux/OS X, it's not _that_ important to address this issue. Especially, when the error message (missing module `semver`) is a standard one and even if the user did not RTFM, he/she should be able to figure out that `semver` has to be installed.
...or have I missed a better way to do this? :D

**Comment by pmualaba** [comment]
This patch works for the calypso login page. Once i login to wordpress.com though, the page is blank with console error: Uncaught TypeError: userUtils.filterUserObject is not a function(anonymous function) @ user.js:128(anonymous function) @ send-request.js:78onload @ index.js:142resolve @ index.js:442onmessage @ index.js:402. I don't have this error on a Mac...

**Comment by dmsnell** [comment]
cc: @alexdresko

**Comment by alexdresko** [comment]
Oh, rad! Thanks for keeping me in the loop, @dmsnell. Totes excited!

**Comment by Viper007Bond** [comment]
Pinging the people who previously participated in tickets about Windows support:

@localpcguy @austinsdev @limweb @HammyHavoc @DigitalDavo @hughc

**Comment by klimeryk** [comment]
@pmualaba: thanks for testing! However, I can't seem to reproduce the error you've reported. I've logged out, visited `http://calypso.localhost:3000/`, got redirected to `http://calypso.localhost:3000/devdocs/start`, logged in with WordPress.com, got redirected to `http://calypso.localhost:3000/devdocs/welcome` and could use everything as usual. No errors in the console.
Can you reproduce the bug in your setup and share the steps needed to do so?

To my fellow Automatticians: thanks for pinging the interested parties :+1:

**Comment by localpcguy** [comment]
Thanks for the efforts on this @klimeryk. I was able to get it running using your directions on a Windows 10 PC.

Guessing there isn't any way, without significant changes like switching build tools, for this to run on the default Windows console (cmd.exe)? Even if we considered extensions to cmd (like GoW, GNU on Windows)? MSYS doesn't intimidate me, but as I've said in other threads, it could inhibit the average WordPress developer who's currently running WordPress under WAMP or similar and doesn't really use any build tools.

**Comment by klimeryk** [comment]
Thank you @localpcguy for testing!
I agree, that it'd be best to get it working without the involvement of MSYS (or Cygwin and the like), just plain `cmd` (however plain and cumbersome it is ;)). The problem is `make` - I can't find an up to date version of it that runs natively on Windows. The only one I found is the [gnuwin32](http://gnuwin32.sourceforge.net/packages/make.htm) and [GNU on Windows](https://github.com/bmatzelle/gow/wiki/executables_list), but both of them are `3.81`, which is over 9 years old at this moment! I _think_ we don't use any of the newer features of `make`, but I don't want the Windows build to be holding us back if ever needed those.

**Comment by liquidboy** [comment]
just tried pulling the 'add/windows-support' branch ... followed the 'docs\windows.md' instructions ...  
and when I get to the point of 'make run'  in a msys console I get the following error : 

$ make run
             _
    ___ __ _| |_   _ _ __  ___  ___
   / __/ _` | | | | | '_ \/ __|/ _ \
  | (_| (_| | | |_| | |_) \__ \ (_) |
   \___\__,_|_|\__, | .__/|___/\___/
               |___/|_|

Makefile:91: recipe for target 'node-version' failed
make: *** [node-version] Error 127

**Comment by klimeryk** [comment]
Thank you for testing, @liquidboy!
* Are you sure node is available?
* What does `node --version` print?
* Have you installed `semver` as per the instructions?

**Comment by liquidboy** [comment]
actually @klimeryk  I know what was wrong, a few weeks ago I installed Windows IoT developer kit which installed Node.js (chakra) v 0.12.7-5  ...  

So my system , in program files (x86) has both 
1. Node.js\   (v 4.2.4)
2. Node.js (chakra)\  (v0.12.7-5)

I solved my problem by temporarily uninstalling Node.js (chakra) .. ill re-install it when I need it (when I do IoT work)..

Worked flaulessly after that  ... thanks again for this effort !

**Comment by klimeryk** [comment]
Great to hear that! :+1:

File: docs/windows.md [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    ```
+    which git
+	/c/Git/cmd/git
+    ```
+9. Install `semver` globally:
+
+    ```
+    npm install semver -g
+    ```
+10. And you're done - you can follow [the normal instructions](https://github.com/Automattic/wp-calypso/blob/master/docs/install.md#installing-and-running) from now on!
```
Comment (gwwar) [inline docs/windows.md]: Are there any steps to generate an ssh key in the msys2 environment? ssh-keygen doesn't seem available.

Comment (klimeryk) [inline docs/windows.md]: You don't need ssh keys to clone a git repository. At least I didn't have to generate any.
I'm guessing `ssh-keygen` should become available when you install ssh, probably through something like `pacman -S ssh` (or `openssh` - can't test right now :/)

Comment (nylen) [inline docs/windows.md]: Our docs recommend using ssh to clone.  PR to amend that to `https:` since it doesn't require this setup - #2278

File: Makefile (hunk 3 above)
Comment (nylen) [inline Makefile]: Given that `node npm` isn't possible on Windows I think this makes sense.

File: Makefile (hunk 4 above)
Comment (nylen) [inline Makefile]: It's not beautiful but it works.  We can always refine that later.

File: Makefile [hunk 7]
```diff
@@ -58,7 +63,12 @@ CALYPSO_ENV ?= $(NODE_ENV)
 
 export NODE_ENV := $(NODE_ENV)
 export CALYPSO_ENV := $(CALYPSO_ENV)
-export NODE_PATH := server:client:.
+export NODE_PATH := server$(SEPARATOR)client$(SEPARATOR).
+ifeq ($(OS),Windows_NT)
+export NPM_GLOBAL_ROOT := $(shell $(NPM) -g root)
```
Comment (nylen) [inline Makefile]: What does `npm -g root` return on Windows that makes this necessary?  Worth documenting in a comment?

File: docs/windows.md [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+
+Please note that native addons via `node-gyp` are [problematic on Windows](https://github.com/nodejs/node-gyp/issues/629), and on MSYS [even more so](https://github.com/nodejs/node-gyp/issues/740).
```
Comment (nylen) [inline docs/windows.md]: Do we have any of these?  I'm guessing none that are problematic, since the app works?

Comment (klimeryk) [inline docs/windows.md]: We have (can't list any from the top of my head, but you can `make distclean` and see what `node-gyp` is building), but they are not required - they just offer better performance.

File: Makefile (hunk 7 above)
Comment (klimeryk) [inline Makefile]: It's just a different path - on Windows it's just in what `npm -g root` returns, while on *nixes you a weird "duplicated" path, like `/usr/lib64/node_modules/npm/node_modules/` (`/usr/lib64/node_modules` is what `npm root -g` returns, and then you have to go again into `node_modules`? :confounded:).
Not sure how to document this - it's just a different path on one platform than the other.

File: docs/windows.md (hunk 6 above)
Comment (klimeryk) [inline docs/windows.md]: Oh, good to know, thanks :+1:

File: docs/windows.md [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+7. Install `make`:
+
+    ```
+    pacman -S make
+    ```
+8. Make sure that `git` is available and the Windows version is used (do *not* install `git` from MSYS2):
+
+    ```
+    which git
+	/c/Git/cmd/git
```
Comment (ebinnion) [inline docs/windows.md]: When I install Git from the link above, and select the following values, my path looks like `/c/Program Files/Git/cmd/git`.

Prompt selections:
screenshot 3
screenshot 4
screenshot 5
screenshot 6
screenshot 7
screenshot 9
screenshot 10

Comment (ebinnion) [inline docs/windows.md]: Did I miss a step when installing Git for Windows? And if so, how can we make that more clear in the doc?

Comment (klimeryk) [inline docs/windows.md]: The first step says:
> When installing, it's best not to use folders with spaces in them.
But you are right that it might need more emphasis. The question is - did it render your install of Git useless? That is, can you use Git for Windows in the MSYS console?

**Comment by nylen** [comment]
This is a really complicated setup.  In fact it would be much easier to just install a Linux VM and run Calypso from there.  But people have asked for it, and you've done a good job finding a combination of software that works pretty well.

I made some tweaks to the documentation in 9f216e6 including screenshots and more information about the optional dependencies that could (will?) fail to install.  I think it's really important to be as clear as possible since there are a lot of potential problems.

In 81d91bc I was also able to remove the requirement to install semver beforehand.  There is a version of semver bundled with the Node.js Windows install, it's just in a different place next to the Node executables.  This is pretty non-intuitive so it needs some documentation in the comments.

After my update, `make node-version` passes `node bin/check-node-version` a MSYS path like `/c/nodejs/node_modules/...`.  But MSYS is smart enough to realize that `node` is a pure Windows executable, and [converts the path to a Windows path](http://www.mingw.org/wiki/Posix_path_conversion):

```
James@Virtualbox MSYS ~/wp-calypso
$ X=/c/git/ node -e 'console.log( process.env.X );'
C:/git/
```

This is also tricky and non-intuitive - perhaps it would be better to write `check-node-version` in a way that it can handle such a path?

One other thing I noticed was a Windows Firewall warning for `node`.  Worth documenting somewhere?

File: Makefile (hunk 2 above)
Comment (nylen) [inline Makefile]: I think a better solution is to make `client/lib/user/shared-utils.js` something like this:

`module.exports = require( 'server/user-bootstrap/shared-utils' );`

**Comment by ebinnion** [comment]
Thank you for all of the work on this @klimeryk and the updated documentation @nylen. Following the new documentation, I was able to successfully get this branch running on my base model Surface 3 (although it did take what seemed like forever to build).

screenshot 16

I can also confirm that I saw the firewall warning about Node JS.

**Comment by klimeryk** [comment]
> I made some tweaks to the documentation in 3a60c01 including screenshots and more information about the optional dependencies that could (will?) fail to install. I think it's really important to be as clear as possible since there are a lot of potential problems.

Awesome! Especially the screenshots are a really good idea :+1: 

> In d835387 I was also able to remove the requirement to install semver beforehand. There is a version of semver bundled with the Node.js Windows install, it's just in a different place next to the Node executables. The fact that npm is set up differently by OS is pretty non-intuitive, so I added some documentation in the comments.

Very nice! I don't know how I've missed that - maybe I got confused, because, as you pointed out, on Windows it's in a different place. Anyway - one less step during the install!

> This is also tricky and non-intuitive - perhaps it would be better to write check-node-version in a way that it can handle an MSYS path?

First of all - I didn't know about the magic MSYS does when handling paths and native Windows executables. The More You Know ™
But I don't understand what's tricky and non-intuitive? The script is handed an absolute path to the `semver` package - the path is correct because of MSYS magic. Is the "magic" part that's tricky and non-intuitive?

About the firewall warning - I think we can omit that in the docs. It highly depends on the setup - if someone has installed a custom firewall, this warning won't be shown (probably the custom firewall will complain instead). And it's a standard Windows prompt - one any user (especially developer) should be familiar with. Nothing unexpected too - node.js is a requirement of Calypso, it's used for package management, so it's natural that it's asking for permission to access the internet.

Thanks you so much @ebinnion and @nylen for testing and improving this patch. Let me know if there's anything else that can be done.

File: Makefile [hunk 10]
```diff
@@ -58,7 +63,20 @@ CALYPSO_ENV ?= $(NODE_ENV)
...
+export NODE_PATH := server$(SEPARATOR)client$(SEPARATOR).
+
+# We use `semver` to check the version of Node.js before installing npm
+# packages or running scripts.  Since this is before npm runs, we need to grab
+# the version of `semver` installed with the npm executable.
+ifeq ($(OS),Windows_NT)
+# On Windows, the npm global install path is under AppData:
+# http://stackoverflow.com/a/26894197
+# `semver` is in the base `node_modules` folder for the `node` installation.
+export SEMVER_GLOBAL_PATH := $(dir $(shell which $(NODE)))/node_modules/npm/node_modules/semver
```
Comment (klimeryk) [inline Makefile]: Good call on the variable name (and adding `semver` to it) - now it's more precise.

File: server/user-bootstrap/shared-utils.js [hunk 11]
```diff
@@ -1,84 +1 @@
...
-			primayBlogUrl = attributes.primary_blog_url || '';
-		return {
-			primarySiteSlug: getSiteSlug( primayBlogUrl ),
-			localeSlug: attributes.language,
-			isRTL: !! ( language && language.rtl )
-		};
-	}
-
-};
+export * from 'client/lib/user/shared-utils';
```
Comment (klimeryk) [inline server/user-bootstrap/shared-utils.js]: I know that omitting `client` would also work here (`lib/user/shared-utils`), but I wanted to make it extra clear what we are doing here ("importing" these shared utils from the client).

**Comment by nylen** [comment]
Tested this out again, including the bits where framework was modified, and looks good to me.

You'll want to test logging in and out on staging after merging this, since the `wpcom_user_bootstrap` config key is disabled in development.

**Comment by klimeryk** [comment]
Deployed - seems nothing exploded or even caught on fire :wink: I did however forget to squash the commits before pushing to master... :persevere:
Sincere thanks to everyone involved, especially @nylen and @ebinnion for the last testing and fixes! :bow:

File: Makefile (hunk 4 above)
Comment (Pomax) [inline Makefile]: there is no reason for semver, the information's already in `process.version`

Comment (aduth) [inline Makefile]: >thankfully there is no reason for using it. The node version is available through `process.version`

`semver` is used by [`bin/check-node-version`](https://github.com/Automattic/wp-calypso/blob/master/bin/check-node-version) for checking that the current Node version satisfies what we've defined as the minimum requirement as `engines.node` in `package.json`, prompting the developer to upgrade if they're using an earlier version.

>semver should not be available by default on any OS (it's an external package)

Which is true, but as a workaround to needing to `npm install semver` before checking the current version, we assume that the global `npm` package [defines it as a dependency](https://github.com/npm/npm/blob/ac9c70d5995a04614a68f554c0e88b5a48ce2858/package.json#L92) and call `semver` from its `node_modules` directory.

If this is enough of a problem, we could probably change this to:

```make
node-version:
	@$(NPM) install semver && $(BIN)/check-node-version
```

... and remove the shenanigans around `SEMVER_GLOBAL_PATH`

Comment (Pomax) [inline Makefile]: right, that's what I mean: check-node-version doesn't need `semver` at all, it is run through node already, and so has access to node's universally accessible `process.version` (without any requires)

Comment (nylen) [inline Makefile]: Loading semver this way is unorthodox, yes, but it is well-explained in the comments and doesn't seem to be causing any problems.  A PR to do a simple version compare would be fine, but we shouldn't use string comparison for this, at the very least we'd want to get rid of the leading `v` and compare each component between `.`s.  Adding in the extra code to do this doesn't seem like a clear win to me over the current approach.

The goal of this check is to quickly execute at the beginning of the build process and abort if continuing is a waste of time.  This should be done before `npm install` since that takes a while due to all of our dependencies.  And as far as `npm install` flagging based on the `engines.node` property...

> Note that, unless the user has set the engine-strict config flag, this field is advisory only. --[npm docs](https://docs.npmjs.com/files/package.json#engines)

(Somewhat ironically, the `make node-version` check uses `process.exitCode` from Node 0.12.x, so it doesn't actually abort the build for older Node versions.  PR to fix this at #3581.)

Comment (Pomax) [inline Makefile]: it is, though, so far five people I work with ran into semver not actually working at all, and the bin/check-node-version script crashing their attempts to run it, which is why I brought this up.

Comment (nylen) [inline Makefile]: I see now that you created an issue for this (#3048) - let's take the conversation over there.
